gpt_index/indices/vector_store/chroma.py

```diff
@@ -0,0 +1,139 @@
+from gpt_index.schema import BaseDocument
+from gpt_index.utils import get_new_id
+
+
+class GPTChromaIndex(BaseGPTIndex[ChromaIndexStruct]):
+    """GPT Chroma Index.
+
+    The GPTChromaIndex is a data structure where nodes are keyed by
+    embeddings, and those embeddings are stored within a Chroma collection.
+    During index construction, the document texts are chunked up,
+    converted to nodes with text; they are then encoded in
+    document embeddings stored within Chroma.
+
+    During query time, the index uses Chroma to query for the top
+    k most similar nodes, and synthesizes an answer from the
+    retrieved nodes.
+
+    Args:
+        text_qa_template (Optional[QuestionAnswerPrompt]): A Question-Answer Prompt
+            (see :ref:`Prompt-Templates`).
+        embed_model (Optional[BaseEmbedding]): Embedding model to use for
+            embedding similarity.
+        chroma_collection (Optional[Any]): Collection instance from `chromadb` package
+
+    """
+
+    index_struct_cls = ChromaIndexStruct
+
+    def __init__(
+        self,
+        documents: Optional[Sequence[DOCUMENTS_INPUT]] = None,
+        index_struct: Optional[ChromaIndexStruct] = None,
+        text_qa_template: Optional[QuestionAnswerPrompt] = None,
+        llm_predictor: Optional[LLMPredictor] = None,
+        embed_model: Optional[BaseEmbedding] = None,
+        chroma_collection: Optional[Any] = None,
+        **kwargs: Any,
+    ) -> None:
+        """Initialize params."""
+        import_err_msg = (
+            "`chromadb` package not found, please run `pip install chromadb`"
+        )
+        try:
+            import chromadb  # noqa: F401
+        except ImportError:
+            raise ValueError(import_err_msg)
+        from chromadb.api.models.Collection import Collection
+
+        self._collection = cast(Collection, chroma_collection)
+
+        self.text_qa_template = text_qa_template or DEFAULT_TEXT_QA_PROMPT
+        super().__init__(
+            documents=documents,
+            index_struct=index_struct,
+            llm_predictor=llm_predictor,
+            embed_model=embed_model,
+            **kwargs,
+        )
+
+    @classmethod
+    def get_query_map(cls) -> Dict[str, Type[BaseGPTIndexQuery]]:
+        """Get query map."""
+        return {
+            QueryMode.DEFAULT: GPTChromaIndexQuery,
+            QueryMode.EMBEDDING: GPTChromaIndexQuery,
+        }
+
+    def _add_document_to_index(
+        self,
+        index_struct: ChromaIndexStruct,
+        document: BaseDocument,
+        text_splitter: TokenTextSplitter,
+    ) -> None:
+        """Add document to index."""
+        if not self._collection:
+            raise ValueError("Collection not initialized")
+        nodes = self._get_nodes_from_document(document, text_splitter)
+
+        embeddings = []
+        metadatas = []
+        ids = []
+        documents = []
+
+        for n in nodes:
+            if n.embedding is None:
```

Comment:
might be worth calling `_get_node_embedding_tups` instead. if you pull upstream `main` branch, you'll see we now abstract out this helper function to make use of embedding batching. see `gpt_index/indices/vector_store/simple.py` as an example